dtensor-keras-bert.py

```diff
@@ -11,320 +11,765 @@
-    dtensor.initialize_tpu_system()
-
-  dprint("Device type:", args.device_type)
-  dprint("Num local devices:", dtensor.num_local_devices(args.device_type))
-  dprint("Num global devices:", dtensor.num_global_devices(args.device_type))
-
-  # Creates the DTensor device mesh.
-  mesh_dims = [(BATCH_DIM, args.batch_dim_size),
-               (MODEL_DIM, args.model_dim_size)]
-  mesh = dtensor.create_distributed_mesh(
-      mesh_dims, device_type=args.device_type)
-
-  # Ensure model replicas are initialized identically by using an identical
-  # RNG seed across the clients (for numpy, tf and keras)
-  tf.keras.utils.set_random_seed(1337)
-  tf.keras.backend.experimental.enable_tf_random_generator()
+    start_step = 0
+    print('start up step: ', start_step)
+    with tf.io.gfile.GFile(step_file_path, 'w') as f:
+      f.write(str(start_step))
+  return start_step
+
+
+# ============================ Tensorboard =======================================
+def config_tensorboard(logging_dir_path):
+  import datetime
+  current_time = datetime.datetime.now().strftime('%Y%m%d-%H%M%S')
+  train_log_dir = os.path.join(logging_dir_path, current_time + 'train')
+  tf.io.gfile.makedirs(train_log_dir)
+  train_summary_writer = tf.summary.create_file_writer(train_log_dir)
+  return train_summary_writer
+
 
-  # Data, model, and optimizer.
-  dataset, pack_fn = get_dataset(mesh)
+def start_trace(logging_dir_path, mesh):
+  dtensor.barrier(mesh)
+  tf.summary.trace_on(graph=True, profiler=False)
+  tf.profiler.experimental.start(logdir=logging_dir_path)
 
-  # FIXME(rainwoodman): The run_on() should have been merged into the
-  # keras.dtensor API points. Without run_on() the eagerly executed code
-  # in Keras uses the default mesh which can be wrong and cause a segfault.
-  with dtensor.run_on(mesh):
-    model = get_model(mesh)
-    optimizer = tf.keras.dtensor.experimental.optimizers.Adam(
-        learning_rate=0.001, mesh=mesh)
 
-  # Train the model
-  train_model(model, optimizer, mesh, dataset, pack_fn)
+def end_trace(mesh):
+  dtensor.barrier(mesh)
+  tf.summary.trace_off()
+  tf.profiler.experimental.stop(save=True)
 
-  # Save a check point
-  cpt = dtensor.DTensorCheckpoint(mesh=mesh, root=model)
-  saved_path = cpt.save(os.path.join(args.prefix, "bert-checkpoint-1/cpt"))
 
-  # Then load it back
-  cpt.restore(saved_path)
+# ============================ Main =======================================
+
+
+def main():
+  args = ap.parse_args()
 
+  print('tensorflow version', tf.__version__)
+
+  # ============================ Cluster and Mesh ============================
+  is_distribute_training = bool(args.global_jobs)
+
+  # CPU device is needed for checkpoint, even when running on GPU mesh
+  # we need to config 1:1 mapping between accelerator and CPU for checkpoint.
+  # This need to happen before we init dtensor multi client.
+  cpu_devices = configure_virtual_devices(args.num_accelerator, 'CPU')
+
+  if is_distribute_training:
```

Comment:
Let's use the environment variable directly instead of overriding them with command-line arguments. I can do this update when moving to initialize_accelerator_systems. (hopefully will land today's nightly)

Reply:
Sure.